src/Files/TemporaryFile.php

```diff
@@ -53,11 +53,13 @@ public function sync(): TemporaryFile
     public function copyFrom($filePath, string $disk = null): TemporaryFile
     {
         if ($filePath instanceof UploadedFile) {
-            $filePath = $filePath->getRealPath();
+            $realPath = $filePath->getRealPath();
+        } else {
+            $realPath = realpath($filePath);
         }
 
-        if ($disk === null && realpath($filePath) !== false) {
-            $readStream = fopen($filePath, 'rb');
+        if ($realPath !== false) {
+            $readStream = fopen($realPath, 'rb');
```

Comment:
This would break my case where `UploadedFile` is stored in S3 and not on the local disk. In this case,`$filePath->getRealPath()` should never return false so it'll always try to read from the local disk.

I think where my PR https://github.com/SpartnerNL/Laravel-Excel/pull/4132 may have gone wrong is that by default `UploadedFile` doesn't use a disk at all. So when you pass `local` as the disk, it tries to read it from the local disk which doesn't have the file. The easiest solution that'd work for both of us is to simply remove the `$disk === null` check from line 59 but I'm not sure if that'd cause an issue for some because if the file existed locally then it'd use that instead of the passed disk. The other option is to check if `($disk === null || $disk === 'local') && realpath($filePath) !== false` but that seems dirty.

Reply:
How about this approach instead?
```php
if ($disk === null && realpath($filePath) !== false) {
    $readStream = fopen($filePath, 'rb');
} else {
    if ($filePath instanceof UploadedFile) {
        $filePath = $filePath->getRealPath();
    }

    $diskInstance = app('filesystem')->disk($disk);

    if ($diskInstance->exists($filePath)) {
        $readStream = $diskInstance->readStream($filePath);
    }
}

if (empty($readStream)) {
    $logPath = '[' . $filePath . ']';

    if ($disk) {
        $logPath .= ' (' . $disk . ')';
    }

    throw new FileNotFoundException('File ' . $logPath . ' does not exist and can therefor not be imported.');
}
```

Reply:
I think this would work for me but does this work for your case? I think the root of the issue is that `UploadedFile` is saved to the `/tmp` directory by default and not in Laravel's `storage` folder (a disk). But with my PR (and I think the example above) the `/tmp` directory would not be checked unless `$disk` is null.

Something I thought of last night was maybe creating a new method that contains the read from disk logic:
```php
protected function getReadStreamFromDisk($filePath, $disk)
{
    $diskInstance = app('filesystem')->disk($disk);

    if (!$diskInstance->exists($filePath)) {
        $logPath = '[' . $filePath . ']';

        if ($disk) {
            $logPath .= ' (' . $disk . ')';
        }

        throw new FileNotFoundException('File ' . $logPath . ' does not exist and can therefor not be imported.');
    }

    return $diskInstance->readStream($filePath);
}
```
And then in `copyFrom` we could do something like this (which could totally be cleaned up):
```php
public function copyFrom($filePath, string $disk = null): TemporaryFile
{
    if ($filePath instanceof UploadedFile) {
        $realPath = $filePath->getRealPath();

        // check if the UploadedFile is on the filesystem first since it's that way by default (and not on a Laravel disk)
        if (realpath($realPath) !== false) {
            $readStream = fopen($realPath, 'rb');
        } else {
            // it's not local so let's check the disk
            $readStream = $this->getReadStreamFromDisk($realPath, $disk);
        }
    } else {
        // if it's not an UploadedFile, the disk is null and it exists on the local filesystem
        if ($disk === null && realpath($filePath) !== false) {
            // read from the filesystem
            $readStream = fopen($filePath, 'rb');
        } else {
            // otherwise read it from the disk
            $readStream = $this->getReadStreamFromDisk($filePath, $disk);
        }
    }

    $this->put($readStream);

    if (is_resource($readStream)) {
        fclose($readStream);
    }

    return $this->sync();
}
```
Does this pass your tests? @patrickbrouwers Thoughts?

Reply:
I think it would be good to get all use cases listed out so we can have unit tests for each. Perhaps extracting that logic to a another class so we can more easily unit test it.

Can someone help list all possible use cases?

Reply:
For my part, the use case is Livewire file uploads that go directly to S3 (or some other disk) and passing the `UploadedFile` to an import. So the `UploadedFile` should be read from the passed disk - https://livewire.laravel.com/docs/uploads#uploading-directly-to-amazon-s3

The other use case for passing an`UploadedFile` would be the default which is reading it from the OS `/tmp` directory.